### ISSUE TITLE
[10.x] Unable to work with queue

### ISSUE THREAD

**Issue by codercms** [vyuldashev/laravel-queue-rabbitmq#299]
- Laravel Version: 6.x
- Package Version: 10.1.3
- RabbitMQ Version: 3.7.16

**Describe the bug**
1. Unable to run `php artisan rabbitmq:consume rabbitmq --queue=...` or ` php artisan queue:work --queue=...` commands without declaring the queue.
1. When using `php artisan rabbitmq:queue-declare` command, there is no way to push any job using pushRaw method of this package, because `queue-declare` command declaring the queue in different way than pushRaw.

**Steps To Reproduce**
Try to run:
```
php artisan rabbitmq:consume rabbitmq --queue=some
```
It will result in error (`PhpAmqpLib\Exception\AMQPProtocolChannelException  : NOT_FOUND - no queue 'some' in vhost '/'`).

Let's declare the queue:
```
php artisan rabbitmq:queue-declare some
```

And run the queue worker again:
```
php artisan rabbitmq:consume rabbitmq --queue=some
```
Now it's running.

Job class:
```php
class Some implements ShouldQueue
{
    public function fire(RabbitMQJob $job, array $payload): void
    {
        $job->release(5);  
    }
}
```

Trying to dispatch a job through RabbitMQ Management:
```json
{"job": "App\\Jobs\\Some", "data": {}}
```

The job execution will result in error (because of "release" method):
```
PRECONDITION_FAILED - inequivalent arg 'x-dead-letter-exchange' for queue 'some' in vhost '/': received the value 'some' of type 'longstr' but current is none
```
And after that worker falls to an infinite loop with this message: `Channel connection is closed.`

Trying to dispatch the same job from CLI:
```php
Queue::pushRaw(json_encode(['job' => 'App\\Jobs\\Some', 'data' => []]), 'some');
```
It will result in error:
```
PhpAmqpLib/Exception/AMQPProtocolChannelException with message 'PRECONDITION_FAILED - inequivalent arg 'x-dead-letter-exchange' for queue 'some' in vhost '/': received the value 'some' of type 'longstr' but current is none'
```

**Expected behavior**
Everything is works like in 9.x

**Additional context**

**Comment by adm-bome** [comment]
@codercms 

Please verify your issue is resolved with the latest commits
`composer require vyuldashev/laravel-queue-rabbitmq:dev-master`

**Comment by codercms** [comment]
@adm-bome I can confirm it works now.
But I found something interesting about failed jobs, they are never removed from the queue.
Steps to reproduce:
```
php artisan rabbitmq:queue-declare some
php artisan rabbitmq:consume --queue=some --tries=1

Push this:
{"job": "Class that doesn't exists in the app", "data": {}}
```

It will be displayed as failed, but it is still "unacked" in the rabbitmq and when queue worker is restarted - failed job is dispatched again.

**Comment by adm-bome** [comment]
@codercms 

**Thanks for checking. And confirming that your issue is resolved.**


Now for the other issue you desribe above.

Is this special case related to te library or to the worker? As i understand your comment: isn't it a good thing that the job is not lost. But processed again?

Which worker/consumer do you use? I think you use the laravel deafult worker class.

If it's worker/consumer related, I think this special case is out of scope of this library. I may be wrong but will investigate this further.

Can you please file a new issue for this special case?

**Comment by codercms** [comment]
@adm-bome I use default Laravel queue worker, but the logic is mixed between Laravel Queue Worker and this library.
Here is stack trace:
```
Target class [Class that doesn't exists in the app] does not exist. {"exception":"[object] (Illuminate\Contracts\Container\BindingResolutionException(code: 0): Target class [Class that doesn't exists in the app] does not exist. at vendor\laravel\framework\src\Illuminate\Container\Container.php:805)
[stacktrace]
#0 vendor\laravel\framework\src\Illuminate\Container\Container.php(681): Illuminate\Container\Container->build('Class that does...')
#1 vendor\laravel\framework\src\Illuminate\Container\Container.php(629): Illuminate\Container\Container->resolve('Class that does...', Array)
#2 vendor\laravel\framework\src\Illuminate\Foundation\Application.php(776): Illuminate\Container\Container->make('Class that does...', Array)
#3 vendor\laravel\framework\src\Illuminate\Queue\Jobs\Job.php(215): Illuminate\Foundation\Application->make('Class that does...')
#4 vendor\laravel\framework\src\Illuminate\Queue\Jobs\Job.php(202): Illuminate\Queue\Jobs\Job->resolve('Class that does...')
#5 vendor\laravel\framework\src\Illuminate\Queue\Jobs\Job.php(182): Illuminate\Queue\Jobs\Job->failed(Object(Illuminate\Contracts\Container\BindingResolutionException))

job is unacked here (it is rejected, even if consumer started with --tries=1)
>>> #6 vendor\vladimir-yuldashev\laravel-queue-rabbitmq\src\Queue\Jobs\RabbitMQJob.php(91): Illuminate\Queue\Jobs\Job->fail(Object(Illuminate\Contracts\Container\BindingResolutionException))

#7 vendor\laravel\framework\src\Illuminate\Queue\Worker.php(482): VladimirYuldashev\LaravelQueueRabbitMQ\Queue\Jobs\RabbitMQJob->fail(Object(Illuminate\Contracts\Container\BindingResolutionException))
#8 vendor\laravel\framework\src\Illuminate\Queue\Worker.php(469): Illuminate\Queue\Worker->failJob(Object(VladimirYuldashev\LaravelQueueRabbitMQ\Queue\Jobs\RabbitMQJob), Object(Illuminate\Contracts\Container\BindingResolutionException))
>>> #9 vendor\vladimir-yuldashev\laravel-queue-rabbitmq\src\Consumer.php(166): Illuminate\Queue\Worker->markJobAsFailedIfWillExceedMaxAttempts('rabbitmq', Object(VladimirYuldashev\LaravelQueueRabbitMQ\Queue\Jobs\RabbitMQJob), 1, Object(Illuminate\Contracts\Container\BindingResolutionException))    
#10 vendor\laravel\framework\src\Illuminate\Queue\Worker.php(399): VladimirYuldashev\LaravelQueueRabbitMQ\Consumer->markJobAsFailedIfWillExceedMaxAttempts('rabbitmq', Object(VladimirYuldashev\LaravelQueueRabbitMQ\Queue\Jobs\RabbitMQJob), 1, Object(Illuminate\Contracts\Container\BindingResolutionException))
>>> #11 vendor\laravel\framework\src\Illuminate\Queue\Worker.php(372): Illuminate\Queue\Worker->handleJobException('rabbitmq', Object(VladimirYuldashev\LaravelQueueRabbitMQ\Queue\Jobs\RabbitMQJob), Object(Illuminate\Queue\WorkerOptions), Object(Illuminate\Contracts\Container\BindingResolutionException))
#12 vendor\laravel\framework\src\Illuminate\Queue\Worker.php(314): Illuminate\Queue\Worker->process('rabbitmq', Object(VladimirYuldashev\LaravelQueueRabbitMQ\Queue\Jobs\RabbitMQJob), Object(Illuminate\Queue\WorkerOptions))
#13 vendor\vladimir-yuldashev\laravel-queue-rabbitmq\src\Consumer.php(99): Illuminate\Queue\Worker->runJob(Object(VladimirYuldashev\LaravelQueueRabbitMQ\Queue\Jobs\RabbitMQJob), 'rabbitmq', Object(Illuminate\Queue\WorkerOptions))
#14 [internal function]: VladimirYuldashev\LaravelQueueRabbitMQ\Consumer->VladimirYuldashev\LaravelQueueRabbitMQ\{closure}(Object(PhpAmqpLib\Message\AMQPMessage))
#15 vendor\php-amqplib\php-amqplib\PhpAmqpLib\Channel\AMQPChannel.php(1043): call_user_func(Object(Closure), Object(PhpAmqpLib\Message\AMQPMessage))
#16 [internal function]: PhpAmqpLib\Channel\AMQPChannel->basic_deliver(Object(PhpAmqpLib\Wire\AMQPReader), Object(PhpAmqpLib\Message\AMQPMessage))
#17 vendor\php-amqplib\php-amqplib\PhpAmqpLib\Channel\AbstractChannel.php(219): call_user_func(Array, Object(PhpAmqpLib\Wire\AMQPReader), Object(PhpAmqpLib\Message\AMQPMessage))
#18 vendor\php-amqplib\php-amqplib\PhpAmqpLib\Channel\AbstractChannel.php(374): PhpAmqpLib\Channel\AbstractChannel->dispatch('60,60', '\x0Ebarcodes_22128...', Object(PhpAmqpLib\Message\AMQPMessage))
#19 vendor\vladimir-yuldashev\laravel-queue-rabbitmq\src\Consumer.php(114): PhpAmqpLib\Channel\AbstractChannel->wait(NULL, true, NULL)
#20 vendor\laravel\framework\src\Illuminate\Queue\Console\WorkCommand.php(112): VladimirYuldashev\LaravelQueueRabbitMQ\Consumer->daemon('rabbitmq', 'handle_barcode_...', Object(Illuminate\Queue\WorkerOptions))
#21 vendor\laravel\framework\src\Illuminate\Queue\Console\WorkCommand.php(96): Illuminate\Queue\Console\WorkCommand->runWorker('rabbitmq', 'handle_barcode_...')
#22 vendor\vladimir-yuldashev\laravel-queue-rabbitmq\src\Console\ConsumeCommand.php(40): Illuminate\Queue\Console\WorkCommand->handle()
#23 [internal function]: VladimirYuldashev\LaravelQueueRabbitMQ\Console\ConsumeCommand->handle()
...
```

Should I open new issue with this problem?

**Comment by adm-bome** [comment]
@codercms 
When the worker decides to mark the job as failed, the message should be removed from rabbitMQ.
The job was added to the database in the failed jobs table. And the message  should be nacked /rejected.

In this case: the worker directly marks the job as failed, because it makes no sense to try something multiple times when the result will be the same every time. (Can't find the class/job to execute).

Isn't this what happens now, with the current code?
What should be the expected behaviour.

> Side node: The queue should not use DLX.

**Comment by adm-bome** [comment]
@codercms 

I did some research, this issue may be a common issue inside the class `Illuminate/Queue/Jobs/Job` on line: 202 and 215. 

When the job is failed the `failed()` method is called. This method tries to resolve the class that does not exists. This can't be done and an exception is thrown.

The code to reject the message `$this->rabbitmq->reject($this);` was never reached.

To solve this in this library I added an extra PR #314
